github/PULL_REQUEST_TEMPLATE: re-think PR testing

Atemu opened the pull request:
The previous template was quite rigid, vague and not used very often.

What we committers actually need is confidence that the PR works as intended and doesn't break anything unexpectedly. Let's instead motivate PR authors aswell as users of the component(s) to give us that confidence.

This hinges on a few assumptions:

- The author has usually already done some testing of their changes
- Most users affected by the change don't know about the PR
- Users of the component(s) who see the PR but aren't familiar with the contribution process don't feel empowered to help review the changes
- Reviewers find it hard to review components they're not familiar with
- People don't know how much just talking a little about the changes helps us committers judge their worthiness

Let's instead write the template in a way that motivates the:

- author to report on their testing; "selling" their PR to us
- author (or users) to tell us how to easily test the PR
- users to come and test the PR; becoming reviewers

Let's also do away with rigid forms. Reviewers need to interpret and read a bunch of things as part of reviewing PRs anyways and this promotes human communication, so simply let people write freeform messages.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

emilazy commented:
I have to say that the template is already quite long and a pain to go through when filing new PRs and (to a lesser extent) reading them, so I would rather see fewer rather than more lines. No objection to the basic idea here, though.

l0b0 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -4,7 +4,55 @@
+- [x] ... <!-- Please provide further checks you did! -->
+
+<!--
+Invite other users of the component(s) you changed to review your PR by [@-mentioning](https://github.blog/news-insights/mention-somebody-they-re-notified/) them or requesting a review if they're part of the NixOS Github org.
```
It might be useful to tell users how to figure out who to ping for changes to existing packages. Hopefully there's a simpler way than `nix eval --expr 'let pkgs = import <nixpkgs> {}; in builtins.toString (builtins.map (maintainer: "@${maintainer.github}") pkgs.PACKAGE.meta.maintainers)' --impure --raw`.

l0b0 reviewed (approved):
Thank you!

Atemu commented:
> the template is already quite long and a pain to go through when filing new PRs and (to a lesser extent) reading them

I forgot to reply to this:

The info which the template intends to ask authors to provide is the sort of information that any reviewer would need to ascertain anyways before merging: Is the intention sound, does the code do what is intended without breaking anything unexpectedly.
  
I think it's faster and easier for everyone involved to ask people to provide such information up-front rather than reviewers needing to manually ask and/or wait for it (directly or indirectly) during the review. If it wasn't clear how much QA was done for instance, I'd have to ask the author to explain. That would require at least two comments and those add up to usually ~2 days of delay.

At least, that is the intention. If you think some pieces of this info are significantly less interesting to reviewers than others, we should explicitly mention that so that PR authors can prioritise.

fricklerhandwerk commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -4,7 +4,55 @@
+Invite other users of the component(s) you changed to review your PR by [@-mentioning](https://github.blog/news-insights/mention-somebody-they-re-notified/) them or requesting a review if they're part of the NixOS Github org.
+When other users of the component test your PR and find it to work as expected, it greatly increases our confidence in your changes.
 -->
+If you (yes, you!) use the component(s) changed in this PR, please test the component(s) built from this PR and report your findings.
```
```suggestion
If you actively use the component(s) changed in this PR, please test the component(s) built from this PR and report your findings.
```

No need to be cheeky, that phrase is easy to read in a condescending tone. Maybe signal emphasis on usage by saying "actively" or something like that?

fricklerhandwerk commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -4,7 +4,55 @@
+
+<!--
+In order for your PR to be merged, it must be reviewed by someone else.
+To increase the amount of people that are able to review your PR, please provide simple and easy to execute steps that reviewers can follow to convince themselves that your changes work as intended.
```
```suggestion
To increase the number of people who are able to review your PR, please provide simple and easy to execute steps that reviewers can follow to convince themselves that your changes work as intended.
```

"Amount" is for uncountable quantities, people are not things

fricklerhandwerk reviewed (approved):
I agree it's an intimidating amount of information to process for contributors, but I also agree it's always worth raising QA standards, and we have to balance who is going to carry the cost of that.

Eventually most of that should be checked in CI and the relevant instructions conveyed via error messages on-demand.

What is currently described as manual tests can naturally expressed as integration tests. Of course that's even more stuff for contributors to learn, but building patterns and infrastructure can help, and as opposed to these last-minute instructions we can teach these skills in dedicated tutorials that can take arbitrary space.

Atemu commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,14 +1,63 @@
+
+- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
+- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
+- [x] I have not tested these changes <!-- Please describe your testing! -->
```
This is intentionally a bit cheeky but it's intended to nudge the PR author to write at least something here, not to make fun of them for not following the template. Please tell me which way you'd have interpreted it.

Atemu commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,14 +1,63 @@
+- "Open the desktop application, open an image file in it via the `File` menu and apply the xyz filter to it. It should look all xyz'd up and not crash anymore."
+- "`nixos-rebuild build-vm` using the following NixOS config (...) and then open localhost:1234 in a browser. You should see a fancy dashboard and no red warning messages."
+-->
+
+
+
```
This whitespace is intentional to invite the PR author to write something here.

Atemu commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,14 +1,63 @@
 
+
+
+If you actively use the component(s) changed in this PR, please test the component(s) built from this PR and report your findings. It really helps a lot.
```
Ideally, we'd have a section in the contributor's manual that describes the basics of how to even test a PR and link it here: How to check out a PR using git, building a package, building Nixpkgs with a local checkout etc.

Atemu committed:
github/PULL_REQUEST_TEMPLATE: re-think PR testing

The previous template was quite rigid, vague and not used very often.

What we committers actually need is confidence that the PR works as intended and
doesn't break anything unexpectedly. Let's instead motivate PR authors aswell as
users of the component(s) to give us that confidence.

This hinges on a few assumptions:

- The author has usually already done some testing of their changes
- Most users affected by the change don't know about the PR
- Users of the component(s) who see the PR but aren't familiar with the
  contribution process don't feel empowered to help review the changes
- Reviewers find it hard to review components they're not familiar with
- People don't know how much just talking a little about the changes helps us
  committers judge their worthiness

Let's instead write the template in a way that motivates the:

- author to report on their testing; "selling" their PR to us
- author and users to tell us how to easily test the PR
- users to come and test the PR; becoming reviewers

Let's also do away with rigid forms. Reviewers need to interpret and read a
bunch of things as part of reviewing PRs anyways and this promotes human
communication, so simply let people write freeform messages.

infinisil commented:
I want to echo the feedback that the PR template is already quite long, and I'm not convinced that making it longer gives us the intended effect: I suspect it will just lead to people reading it even less because it's _too_ long.  In particular also because not all the information is relevant for each PR, and it's hard to navigate and read in the PR editor. Furthermore it duplicates information in other contributing docs, making it harder to keep it in sync. I suggest reading the discussions in [this recent Lobsters discussion](https://lobste.rs/s/bjix2b/undefeated_pull_request_template) where some of the same sentiment has been shared.

My vision with contributor documentation after https://github.com/NixOS/nixpkgs/pull/245243 (where I rectified part of contributor docs being duplicated and out of sync) is that:
- The PR template be stripped down to a very generic minimal outline along the lines of
  > - [ ] Followed the [generic guidelines](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#conventions)
  > - [ ] Followed the [change-specific guidelines](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#overview)
  > - [x] This change is user-facing `<!-- Untick and remove the subpoints if it's not -->`
  >   - [ ] Breaking change?
  >   - [ ] Changelog written?
  >   - [ ] Documentation updated?
  > - [ ] Tested how?

- All contributor docs be rendered in an easily-searchable and cross-linked way, such that we can add a dedicated link to each checkbox, which then points users to specific instructions based on what they change. So e.g. the "Tested how?" checkpoint would link to a markdown doc like:
  > Different parts of Nixpkgs are tested differently:
  > - `pkgs`: See [here](https://github.com/NixOS/nixpkgs/tree/master/pkgs#package-tests)
  > - `nixos`: See [here](https://github.com/NixOS/nixpkgs/tree/master/nixos#testing-changes)
  > - `lib`: See [here](https://github.com/NixOS/nixpkgs/tree/master/lib#running-tests)
  > - ...

  A generic start is [the overview in `CONTRIBUTING.md`](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#overview).

This way, the goal stops becoming "making the best tradeoff between useful information and template length", but rather "how to organise contributor docs in the most efficient way".

llakala commented:
Working on an alternate concept in #425516. If we end up preferring this version, that's totally fine - I'd just like to get *some* improvement over the current template.
